### ISSUE TITLE
Improved custom enum support

### ISSUE THREAD

**Issue by jedevc** [dagger/dagger#8671]
Opening an issue from https://discord.com/channels/707636530424053791/1123687185837740053/1290092449183698965 - me and @helderco had a chat about this.

Current enum support has a few issues, which we can fix!

- Naming clashes are very possible in go
	- Enum values aren't namespaced at all.
	- Fixed by https://github.com/dagger/dagger/pull/8669

- Some core enum types have stuttering names
	- e.g. `TypeDefKind.STRING_KIND` should be `TypeDefKind.STRING`, `ModuleSourceKind.GIT_SOURCE` should be `ModuleSourceKind.GIT`

- We confuse enum names and values, e.g. in the following go declaration: `const MyEnumValue MyEnum = "foobar"`, the graphql `enumValues.name` is populated as `foobar` - this is weird, because *everything* else is parsed by identifier (functions, args, structs, etc).
	- This is caused by the fact that graphql doesn't have separate names/values for enums, so we picked values.
	- However, with #8498 incoming, we can add custom directives :smile: 
	- So, we can switch the `enumValues` `name` to be derived from the name - and use a custom `@value` directive for the value.
	- This does a few things:
		- The name generation is now internally consistent
		- Names are correctly shared between different modules
		- Values don't need to conform to the graphql naming format (which only allows *very* specific formats)
		- Enum values could be of other types, not just strings!

**Comment by sagikazarmark** [comment]
Enums in Go is a giant PITA. They aren't just badly supported by the language, there are also competing practices out there with varying UX.

Here is how I define enums these days in Dagger modules:

```
// module: trivy
type ReportFormat string

const (
	ReportFormatTable      ReportFormat = "table"
	ReportFormatJSON       ReportFormat = "json"
	ReportFormatTemplate   ReportFormat = "template"
	ReportFormatSARIF      ReportFormat = "sarif"
	ReportFormatCycloneDX  ReportFormat = "cyclonedx"
	ReportFormatSPDX       ReportFormat = "spdx"
	ReportFormatSPDXJSON   ReportFormat = "spdx_json"
	ReportFormatGitHub     ReportFormat = "github"
	ReportFormatCosignVuln ReportFormat = "cosign_vuln"
)
```

I would expect to be able to use a _similar_ version of this on the other side: probably `dagger.TrivyReportFormatMarkdown`?

I chose to prefix the enum values manually because there might be an overlap between different kinds of enums.

**Comment by jedevc** [comment]
> I chose to prefix the enum values manually because there might be an overlap between different kinds of enums.

Yup, allowing prefixing sounds good to me, we can do this without too much trouble, I'll add this to the list of things to do.

---

In terms of picking the enum values, with the third point in my list, that would completely detach names from values, so you're free to pick the values *however* you like - even including non-name characters like `{[#@`, etc.

**Comment by sagikazarmark** [comment]
> > I chose to prefix the enum values manually because there might be an overlap between different kinds of enums.
> 
> Yup, allowing prefixing sounds good to me, we can do this without too much trouble, I'll add this to the list of things to do.

What I meant here was actually not doing any kind of magic in the code generator, but encouraging users to prefix their enums themselves. That was they can account for any collisions within the module and the generated code will stay relatively close to what's in the module.


> In terms of picking the enum values, with the third point in my list, that would completely detach names from values, so you're free to pick the values _however_ you like - even including non-name characters like `{[#@`, etc.

I think that's a good idea.

**Comment by jedevc** [comment]
> What I meant here was actually not doing any kind of magic in the code generator, but encouraging users to prefix their enums themselves. That was they can account for any collisions within the module and the generated code will stay relatively close to what's in the module.

Sadly, I think some simple level of magic is needed - e.g. `const MyEnumValue MyEnum = "value"` should appear as `MyEnum.Value` in the graphql API. Without that, the python/typescript/etc will have very weird stuttery names like `MyEnum.MyEnumValue`. Go's enums kinda suck, but we can at least keep them in their own pool of sadness and not spread it into the other SDKs.

**Comment by nipuna-perera** [comment]
This covers pretty much everything I want from ENUMS. It will be a big breaking change or folks using ENUMS today though. But I think this is necessary.

**Comment by sagikazarmark** [comment]
> > What I meant here was actually not doing any kind of magic in the code generator, but encouraging users to prefix their enums themselves. That was they can account for any collisions within the module and the generated code will stay relatively close to what's in the module.
> 
> Sadly, I think some simple level of magic is needed - e.g. `const MyEnumValue MyEnum = "value"` should appear as `MyEnum.Value` in the graphql API. Without that, the python/typescript/etc will have very weird stuttery names like `MyEnum.MyEnumValue`. Go's enums kinda suck, but we can at least keep them in their own pool of sadness and not spread it into the other SDKs.

A couple ideas:

1. as a convention, Go enums MUST have the type name as a prefix, so you know what to strip from their names
2. use pragmas to redefine the name (needs validation though to make sure there are no collisions)
3. come up with an alternative enum syntax for Go (less idiomatic for Go developers, but it's probably closer to python's and typescript's enum semantic)

The challenge here is module developers may have to prefix values manually, to avoid collisions within a module, but that may result in equally stuttery names if Dagger also does some magic prefixing.

Sticking with the above example:

```
// module: trivy
type ReportFormat string

const (
	ReportFormatTable      ReportFormat = "table"
	ReportFormatJSON       ReportFormat = "json"
)
```

If it becomes `dagger.TrivyReportFormatReportFormatTable`, it's probably not ideal.

I would probably go with option 1 from the above list for now. That limitation can always be lifted later, if necessary.

**Comment by charjr** [comment]
Just adding a use-case, In PHP there are currently three kinds of enum:
1. unit enum: names without values
2. string-backed enum: names with string values
3. integer-backed enum: names with integer values

If the PHP SDK implements enum support using values, we can only support 2. 
If I can rely on names, _all enums_ can be supported in a consistent manner.

**Comment by jedevc** [comment]
These aren't all going to make it into v0.14.0 - moving out of the milestone.
